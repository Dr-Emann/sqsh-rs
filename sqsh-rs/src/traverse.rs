use crate::{error, Archive, File, FileType};
use bstr::BStr;
use sqsh_sys as ffi;
use std::fmt;
use std::iter::FusedIterator;
use std::marker::PhantomData;
use std::ptr::NonNull;

/// An efficient traversal of the archive.
///
/// This is an efficient but low-level interface to the archive.
///
/// This acts as a lending iterator (the entries borrow from the iterator), and so can't actually
/// implement the `Iterator` trait.
pub struct Traversal<'archive> {
    inner: NonNull<ffi::SqshTreeTraversal>,
    _marker: PhantomData<&'archive Archive<'archive>>,
}

#[derive(Copy, Clone)]
pub struct Entry<'traversal, 'archive> {
    inner: &'traversal ffi::SqshTreeTraversal,
    _marker: PhantomData<&'traversal Traversal<'archive>>,
}

#[derive(Copy, Clone)]
pub struct Path<'traversal> {
    entry: Entry<'traversal, 'traversal>,
}

/// An iterator over the segments of a path.
#[derive(Debug, Clone)]
pub struct PathSegments<'traversal> {
    inner: &'traversal ffi::SqshTreeTraversal,
    depth: usize,
    index: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum State {
    /// The traversal iterator is currently pointing at an object it will visit only once.
    ///
    /// This includes regular files, special files like named pipes, symlinks, or devices, and
    /// directories that will not be descended into because of a configured max depth.
    Normal,
    /// The traversal is currently at a directory entry, the traversal will enter the directory
    DirectoryFirst,
    /// The traversal is currently at a directory entry, the traversal just exited the directory
    ///
    /// It may be desirable to skip this entry, since it was visited in [`Self::DirectoryFirst`]
    /// already.
    DirectorySecond,
}

impl State {
    pub fn is_second_visit(self) -> bool {
        self == Self::DirectorySecond
    }
}

impl<'archive> Traversal<'archive> {
    pub(crate) unsafe fn new(inner: NonNull<ffi::SqshTreeTraversal>) -> Self {
        Self {
            inner,
            _marker: PhantomData,
        }
    }

    pub fn set_max_depth(&mut self, max_depth: usize) {
        unsafe { ffi::sqsh_tree_traversal_set_max_depth(self.inner.as_ptr(), max_depth) }
    }

    /// Attempt to advance the traversal to the next entry.
    pub fn advance(&mut self) -> error::Result<Option<Entry<'_, 'archive>>> {
        let mut err = 0;
        let has_next = unsafe { ffi::sqsh_tree_traversal_next(self.inner.as_ptr(), &mut err) };
        if err != 0 {
            return Err(error::new(err));
        }
        Ok(has_next.then_some(Entry {
            inner: unsafe { self.inner.as_ref() },
            _marker: PhantomData,
        }))
    }
}

impl Drop for Traversal<'_> {
    fn drop(&mut self) {
        unsafe {
            ffi::sqsh_tree_traversal_free(self.inner.as_ptr());
        }
    }
}

impl<'traversal, 'archive> Entry<'traversal, 'archive> {
    /// The depth of this entry.
    ///
    /// The root entry has a depth of 0.
    pub fn depth(self) -> usize {
        unsafe { ffi::sqsh_tree_traversal_depth(self.inner) }
    }

    /// The name of this entry.
    ///
    /// This is the same as the last [path segment][Self::path_segments], or an empty string if
    /// there are no path segments.
    ///
    /// The root entry has an empty name.
    pub fn name(self) -> &'traversal BStr {
        let mut len = 0;
        let ptr = unsafe { ffi::sqsh_tree_traversal_name(self.inner, &mut len) };
        let slice = unsafe { std::slice::from_raw_parts(ptr.cast::<u8>(), len) };
        BStr::new(slice)
    }

    /// The path of this entry.
    ///
    /// This will be the relative path from the root of the traversal to this entry.
    pub fn path(self) -> Path<'traversal> {
        Path::new(self)
    }

    /// Open the current entry.
    pub fn open(self) -> error::Result<File<'archive>> {
        let mut err = 0;
        let file = unsafe { ffi::sqsh_tree_traversal_open_file(self.inner, &mut err) };
        let file = match NonNull::new(file) {
            Some(file) => file,
            None => return Err(error::new(err)),
        };
        Ok(unsafe { File::new(file) })
    }

    pub fn file_type(self) -> FileType {
        let file_type = unsafe { ffi::sqsh_tree_traversal_type(self.inner) };
        FileType::try_from(file_type).unwrap()
    }

    /// The directory entry for this entry. This will be present for everything but the root entry.
    pub fn directory_entry(self) -> Option<crate::directory::DirectoryEntry<'traversal, 'archive>> {
        let iterator = unsafe { ffi::sqsh_tree_traversal_iterator(self.inner) };
        if iterator.is_null() {
            return None;
        }
        Some(unsafe { crate::directory::DirectoryEntry::new(&*iterator) })
    }

    pub fn state(self) -> State {
        let state = unsafe { ffi::sqsh_tree_traversal_state(self.inner) };
        match state {
            ffi::SqshTreeTraversalState::SQSH_TREE_TRAVERSAL_STATE_INIT
            | ffi::SqshTreeTraversalState::SQSH_TREE_TRAVERSAL_STATE_FILE => State::Normal,
            ffi::SqshTreeTraversalState::SQSH_TREE_TRAVERSAL_STATE_DIRECTORY_BEGIN => {
                State::DirectoryFirst
            }
            ffi::SqshTreeTraversalState::SQSH_TREE_TRAVERSAL_STATE_DIRECTORY_END => {
                State::DirectorySecond
            }
            _ => {
                debug_assert!(false, "unexpected state");
                State::Normal
            }
        }
    }
}

impl fmt::Debug for Entry<'_, '_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("Entry")
            .field("state", &self.state())
            .field("depth", &self.depth())
            .field("path", &self.path())
            .field("file_type", &self.file_type())
            .finish_non_exhaustive()
    }
}

impl<'traversal> Path<'traversal> {
    pub(crate) fn new(entry: Entry<'traversal, '_>) -> Self {
        Self { entry }
    }

    /// The path segments of the current entry.
    ///
    /// This is an iterator over the segments of the path (relative to the root of the traversal)
    ///
    /// For example, if the current entry is `foo/bar/baz`,
    /// the path segments are `["foo", "bar", "baz"]`.
    pub fn segments(self) -> PathSegments<'traversal> {
        PathSegments::new(self.entry)
    }
}

impl fmt::Debug for Path<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self.to_string())
    }
}

impl fmt::Display for Path<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut segments = self.segments();
        if let Some(segment) = segments.next() {
            write!(f, "{}", segment)?;
            for segment in segments {
                write!(f, "/{}", segment)?;
            }
        }
        Ok(())
    }
}

impl<'a> PathSegments<'a> {
    pub(crate) fn new(entry: Entry<'a, '_>) -> Self {
        Self {
            inner: entry.inner,
            depth: entry.depth(),
            index: 0,
        }
    }

    fn segment(&self, index: usize) -> &'a BStr {
        let mut len = 0;
        let segment = unsafe { ffi::sqsh_tree_traversal_path_segment(self.inner, &mut len, index) };

        debug_assert!(!segment.is_null());
        let slice = unsafe { std::slice::from_raw_parts(segment.cast::<u8>(), len) };
        BStr::new(slice)
    }
}

impl<'a> Iterator for PathSegments<'a> {
    type Item = &'a BStr;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index >= self.depth {
            return None;
        }

        let result = self.segment(self.index);
        self.index += 1;
        Some(result)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.len(), Some(self.len()))
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.index = self.depth.min(self.index.saturating_add(n));
        self.next()
    }
}

impl ExactSizeIterator for PathSegments<'_> {
    fn len(&self) -> usize {
        self.depth - self.index
    }
}

impl FusedIterator for PathSegments<'_> {}

impl<'a> DoubleEndedIterator for PathSegments<'a> {
    fn next_back(&mut self) -> Option<Self::Item> {
        if self.index >= self.depth {
            return None;
        }
        self.depth -= 1;
        Some(self.segment(self.depth))
    }

    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
        self.depth = self.index.max(self.depth.saturating_sub(n));
        self.next_back()
    }
}
