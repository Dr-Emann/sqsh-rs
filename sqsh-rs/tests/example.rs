use bstr::BString;
use sqsh_rs::traverse::Traversal;
use sqsh_rs::{Archive, DirectoryIterator, FileType, Permissions};
use std::fmt::Write;
use std::io::{BufRead, Read};

mod all;

const ARCHIVE_PATH: &str = "tests/data/test.sqsh";

fn archive() -> Archive<'static> {
    Archive::new(ARCHIVE_PATH).unwrap()
}

#[test]
fn open_archive() {
    let _archive = archive();
}

#[test]
fn mem_open_archive() {
    let data = std::fs::read(ARCHIVE_PATH).unwrap();
    let _archive = Archive::from_slice(&data).unwrap();
}

#[test]
fn superblock() {
    let archive = archive();
    let superblock = archive.superblock();
    insta::assert_debug_snapshot!(superblock);
}

#[test]
fn easy_contents_not_exists() {
    let archive = archive();
    let err = archive.read("not_exists").unwrap_err();
    assert_eq!(err.io_error_kind(), std::io::ErrorKind::NotFound);
    insta::assert_snapshot!(err, @"No such file or directory");
}

#[test]
fn easy_contents_empty() {
    let archive = archive();
    let data = archive.read("empty.file").unwrap();
    assert!(data.is_empty());
}

#[test]
fn easy_contents_one() {
    let archive = archive();
    let data = archive.read("one.file").unwrap();
    assert_eq!(data.len(), 1);
    assert_eq!(data, "a".as_bytes());
}

#[test]
fn easy_contents_of_directory() {
    let archive = archive();
    let err = archive.read("subdir").unwrap_err();
    assert_eq!(err.io_error_kind(), std::io::ErrorKind::Other);
    insta::assert_snapshot!(err, @"Not a file");
}

#[test]
fn easy_permissions() {
    let archive = archive();
    let permissions = archive.permissions("one.file").unwrap();
    assert_eq!(
        permissions,
        Permissions::UserRW | Permissions::GroupRead | Permissions::OtherRead
    );
}

#[test]
fn open_file() {
    let archive = archive();
    let file = archive.open("one.file").unwrap();
    assert!(!file.is_extended());
}

#[test]
fn open_dir() {
    let archive = archive();
    let dir = archive.open("subdir").unwrap();
    assert_eq!(dir.file_type(), Some(sqsh_rs::FileType::Directory));
    insta::assert_debug_snapshot!("subdir debug", dir);
}

#[test]
fn reopen_by_id() {
    let archive = archive();
    let file1 = archive.open("one.file").unwrap();
    let inode_ref = file1.inode_ref();
    let file2 = archive.open_ref(inode_ref).unwrap();
    assert_eq!(format!("{file1:?}"), format!("{file2:?}"));
}

#[test]
fn reader_read_by_byte() {
    let archive = archive();
    let file = archive.open("short.file").unwrap();
    let mut reader = file.reader().unwrap();
    let mut buf = [0u8; 1];
    for i in 0..4 {
        let n = reader.read(&mut buf).unwrap();
        assert_eq!(n, 1);
        let expected = match i {
            0 => b'a',
            1 => b'b',
            2 => b'c',
            3 => b'\n',
            _ => unreachable!(),
        };
        assert_eq!(buf[0], expected);
    }
    let n = reader.read(&mut buf).unwrap();
    assert_eq!(n, 0);
}

#[test]
fn reader_buf_read() {
    let archive = archive();
    let file = archive.open("1MiB.file").unwrap();
    let mut reader = file.reader().unwrap();
    let mut total_size = 0;
    loop {
        let buf = reader.fill_buf().unwrap();
        if buf.is_empty() {
            break;
        }
        let len = buf.len();
        assert!(buf.iter().all(|&b| b == b'A'));
        total_size += buf.len();

        // Only consume part of the buffer, ensure that the remaining part is returned by fill_buf
        reader.consume(2);
        let remaining_buf = reader.fill_buf().unwrap();
        assert_eq!(remaining_buf.len(), len - 2);
        reader.consume(len - 2);
    }
    assert_eq!(total_size, 1024 * 1024);
}

#[test]
fn resolver() {
    let archive = archive();
    let mut resolver = archive.path_resolver().unwrap();

    assert_eq!(resolver.current_name(), None);
    assert_eq!(resolver.current_file_type(), Some(FileType::Directory));

    let current = resolver.open().unwrap();
    let root = archive.open("").unwrap();
    assert_eq!(format!("{current:?}"), format!("{root:?}"));

    resolver.advance().unwrap();

    assert_eq!(resolver.current_name(), Some(bstr::BStr::new(b"1MiB.file")));
    assert_eq!(resolver.current_file_type(), Some(FileType::File));

    let current = resolver.open().unwrap();
    assert_eq!(current.size(), 1024 * 1024);
}

#[test]
fn walk_whole_dir() {
    let archive = archive();
    let mut path_resolver = archive.path_resolver().unwrap();

    let mut names = Vec::new();
    while path_resolver.advance().unwrap() {
        names.push(
            std::str::from_utf8(path_resolver.current_name().unwrap())
                .unwrap()
                .to_owned(),
        );
    }

    insta::assert_debug_snapshot!(names);
}

#[test]
fn recursive_directory() {
    #[derive(Debug)]
    #[allow(dead_code)]
    enum Tree {
        File(BString),
        Directory(BString, Vec<Tree>),
    }
    fn visit_directory(mut iter: DirectoryIterator) -> Vec<Tree> {
        let mut trees = Vec::new();
        while let Some(entry) = iter.advance().unwrap() {
            let name = entry.name();
            let tree = if entry.file_type() == Some(FileType::Directory) {
                let dir = entry.open().unwrap();
                let children = visit_directory(dir.as_dir().unwrap());
                Tree::Directory(name.into(), children)
            } else {
                Tree::File(name.into())
            };
            trees.push(tree);
        }
        trees
    }
    let archive = archive();
    let root = archive.root().unwrap();
    let trees = visit_directory(root.as_dir().unwrap());
    insta::assert_debug_snapshot!(trees);
}

#[test]
fn skip_to_last_byte() {
    let archive = archive();
    let file = archive.open("1MiB.file").unwrap();
    assert_eq!(file.size(), 1024 * 1024);

    let mut reader = file.reader().unwrap();
    assert!(reader.block_size() < 1024 * 1024);

    let mut buf = [0u8; 2];

    // Read a partial block in the front
    assert_eq!(reader.read(&mut buf).unwrap(), 2);
    assert_eq!(buf, [b'A', b'A']);

    // Then skip ahead to the last byte
    reader.skip(1024 * 1024 - buf.len() as u64 - 1).unwrap();
    let n = reader.read(&mut buf).unwrap();
    assert_eq!(n, 1);
    assert_eq!(buf[0], b'A');
}

#[test]
fn skip_past_end() {
    let archive = archive();
    let file = archive.open("1MiB.file").unwrap();
    assert_eq!(file.size(), 1024 * 1024);

    let mut reader = file.reader().unwrap();

    assert_eq!(
        reader.skip(1024 * 1024 + 1).unwrap_err().to_string(),
        "Out of bounds"
    );
}

#[test]
fn compression_options() {
    let archive = archive();
    insta::assert_debug_snapshot!(archive.compression_options());
}

fn traversal_str(traversal: &mut Traversal) -> String {
    let mut result = String::new();
    while let Some(entry) = traversal.advance().unwrap() {
        writeln!(
            result,
            "/{} {:?} {}\n{:?}",
            entry.path(),
            entry.state(),
            entry.depth(),
            entry.directory_entry(),
        )
        .unwrap();

        assert_eq!(
            entry.path().segments().next_back().unwrap_or_default(),
            entry.name(),
        );
    }
    result
}

#[test]
fn traverse() {
    let archive = archive();
    let root = archive.root().unwrap();
    let mut traversal = root.traversal().unwrap();
    insta::assert_snapshot!(traversal_str(&mut traversal));
}

#[test]
fn traverse_start_subdir() {
    let archive = archive();
    let subdir = archive.open("subdir").unwrap();

    let mut traversal = subdir.traversal().unwrap();
    insta::assert_snapshot!(traversal_str(&mut traversal));
}

#[test]
fn traverse_start_file() {
    let archive = archive();
    let file = archive.open("/one.file").unwrap();

    let mut traversal = file.traversal().unwrap();
    insta::assert_snapshot!(traversal_str(&mut traversal));
}

#[test]
fn traverse_max_depth() {
    let archive = archive();
    let root = archive.root().unwrap();
    let mut traversal = root.traversal().unwrap();
    traversal.set_max_depth(1);

    insta::assert_snapshot!(traversal_str(&mut traversal));
}
