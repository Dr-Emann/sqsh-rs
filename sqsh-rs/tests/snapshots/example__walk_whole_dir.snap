---
source: sqsh-rs/tests/example.rs
expression: names
---
[
    "1MiB.file",
    "broken.link",
    "deep",
    "dev",
    "empty.file",
    "empty_dir",
    "fifo",
    "one.file",
    "short.file",
    "short.link",
    "socket",
    "socket2",
    "subdir",
]
